File: examples/python/xauth.py
import requests
import json
import uuid
import base64
import hashlib
import subprocess
import platform
from Crypto.Cipher import AES

def get_hwid():
    """Generates a reliable hardware ID (cross-platform)"""
    try:
        if platform.system() == "Windows":
            output = subprocess.check_output('wmic csproduct get uuid', shell=True).decode()
            return output.split('\n')[1].strip()
        elif platform.system() == "Linux":
            with open("/etc/machine-id", "r") as f:
                return f.read().strip()
        elif platform.system() == "Darwin":
            output = subprocess.check_output("ioreg -rd1 -c IOPlatformExpertDevice | awk '/IOPlatformUUID/ { print $3; }'", shell=True).decode()
            return output.strip().strip('"')
    except Exception:
        pass
    # Fallback to mac address if all else fails
    return str(uuid.getnode())


class XAuth:
    """XAuth Client SDK for Python"""
    def __init__(self, app_id, app_secret, base_url="http://localhost:3310"):
        self.app_id = app_id
        self.app_secret = app_secret
        self.base_url = base_url.rstrip("/")
        self.session_id = None
        self.nonce = None

    def initialize(self):
        """Step 1: Initialize the session to get a nonce"""
        payload = {"app_id": self.app_id}
        try:
            response = requests.post(f"{self.base_url}/api/v1/client/initialize", json=payload)
            if response.status_code == 200:
                data = response.json()
                self.session_id = data.get("session_id")
                self.nonce = data.get("nonce")
                return True, "Session initialized"
            return False, response.text
        except Exception as e:
            return False, str(e)

    def _decrypt_response(self, encrypted_data):
        try:
            # Format from server: IV:AuthTag:EncryptedData (all base64)
            iv_b64, tag_b64, data_b64 = encrypted_data.split(":")
            
            iv = base64.b64decode(iv_b64)
            tag = base64.b64decode(tag_b64)
            ciphertext = base64.b64decode(data_b64)
            
            # The session nonce is appended to the secret to form the true encryption key
            raw_key = (self.app_secret + self.nonce).encode('utf-8')
            key = hashlib.sha256(raw_key).digest()

            cipher = AES.new(key, AES.MODE_GCM, nonce=iv)
            decrypted = cipher.decrypt_and_verify(ciphertext, tag)
            
            return json.loads(decrypted.decode('utf-8'))
        except Exception:
            return None

    def validate_license(self, license_key):
        """
        Validates the license key against the XAuth Omega Infrastructure.
        Returns a dict: {'success': bool, 'message'?: str, 'expiry'?: str, 'broadcast'?: str, 'variables'?: dict}
        """
        # Ensure session is initialized
        if not self.session_id or not self.nonce:
            success, msg = self.initialize()
            if not success:
                return {"success": False, "message": f"Initialization failed: {msg}"}

        hwid = get_hwid()

        payload = {
            "license_key": license_key,
            "hwid": hwid,
            "app_secret": self.app_secret,
            "session_id": self.session_id
        }

        try:
            response = requests.post(
                f"{self.base_url}/api/v1/client/validate",
                json=payload
            )
            
            if response.status_code == 200:
                data = response.json()
                # Server returns { data: "encrypted_string" }
                encrypted_payload = data.get("data")
                
                decrypted_data = self._decrypt_response(encrypted_payload)
                if decrypted_data and decrypted_data.get("status") == "success":
                    return {
                        "success": True,
                        "expiry": decrypted_data.get("expiry"),
                        "broadcast": decrypted_data.get("broadcast"),
                        "variables": decrypted_data.get("variables", {})
                    }
                else:
                    return {"success": False, "message": "Failed to verify secure response."}
            else:
                try:
                    error_data = response.json()
                    return {"success": False, "message": error_data.get("message", "Validation rejected")}
                except Exception:
                    return {"success": False, "message": response.text}
        except Exception as e:
            return {"success": False, "message": f"Network Error: {str(e)}"}

    def check_version(self, current_version=None, channel="stable"):
        """
        Securely checks for application updates.
        Returns a dict: {'success': bool, 'version'?: str, 'update_available'?: bool, ...}
        """
        payload = {
            "app_id": self.app_id,
            "app_secret": self.app_secret,
            "channel": channel,
            "current_version": current_version
        }

        try:
            response = requests.post(
                f"{self.base_url}/api/v1/client/verify-version",
                json=payload
            )
            
            if response.status_code == 200:
                data = response.json()
                return {
                    "success": True,
                    "version": data.get("version"),
                    "channel": data.get("channel"),
                    "url": data.get("url"),
                    "checksum": data.get("checksum"),
                    "published_at": data.get("published_at"),
                    "update_available": data.get("update_available"),
                    "broadcast": data.get("broadcast")
                }
            else:
                try:
                    error_data = response.json()
                    return {"success": False, "message": error_data.get("message", "Version check failed")}
                except Exception:
                    return {"success": False, "message": response.text}
        except Exception as e:
            return {"success": False, "message": f"Network Error: {str(e)}"}

    def get_latest_release(self, channel="stable"):
        """
        Publicly checks for the latest application release without a secret.
        Returns a dict: {'success': bool, 'version'?: str, ...}
        """
        try:
            response = requests.get(f"{self.base_url}/api/update/{self.app_id}/{channel}")
            if response.status_code == 200:
                data = response.json()
                return {
                    "success": True,
                    "version": data.get("version"),
                    "channel": data.get("channel"),
                    "url": data.get("url"),
                    "checksum": data.get("checksum"),
                    "published_at": data.get("published_at")
                }
            return {"success": False, "message": response.text}
        except Exception as e:
            return {"success": False, "message": str(e)}


# Example Usage
if __name__ == "__main__":
    # Initialize XAuth
    # In production, use your actual App ID and Secret Key
    auth = XAuth(app_id=1, app_secret="your_app_secret_here")

    print("--- XAuth Omega Python SDK Demo ---")

    # 1. Public Update Check (GET)
    print("\n[1] Checking for updates (Public)...")
    release = auth.get_latest_release(channel="stable")
    if release["success"]:
        print(f"Latest Version: {release['version']} ({release['channel']})")
        print(f"Download URL: {release['url']}")
    else:
        print(f"Update check failed: {release['message']}")

    # 2. Secure Version Verification (POST)
    print("\n[2] Verifying local version (Secure)...")
    ver = auth.check_version(current_version="1.0.0", channel="stable")
    if ver["success"]:
        if ver["update_available"]:
            print(f"A new version {ver['version']} is available!")
        else:
            print("You are running the latest version.")
        
        if ver["broadcast"]:
            print(f"Server Message: {ver['broadcast']}")
    else:
        print(f"Secure check failed: {ver['message']}")

    # 3. License Validation
    print("\n[3] Validating license...")
    license_key = "XXXX-XXXX-XXXX-XXXX" # Replace with a real key
    result = auth.validate_license(license_key)
    
    if result["success"]:
        print(f"License Valid! Expires: {result['expiry']}")
        if result["broadcast"]:
            print(f"Broadcast: {result['broadcast']}")
        print(f"Variables: {result['variables']}")
    else:
        print(f"Validation failed: {result['message']}")

File: examples/nodejs/xauth.js
const crypto = require('crypto');
const os = require('os');
const { execSync } = require('child_process');

function getHwid() {
    try {
        if (os.platform() === 'win32') {
            return execSync('wmic csproduct get uuid', { encoding: 'utf8' }).split('\n')[1].trim();
        } else if (os.platform() === 'linux') {
            return require('fs').readFileSync('/etc/machine-id', 'utf8').trim();
        } else if (os.platform() === 'darwin') {
            const output = execSync("ioreg -rd1 -c IOPlatformExpertDevice | awk '/IOPlatformUUID/ { print $3; }'", { encoding: 'utf8' });
            return output.replace(/"/g, '').trim();
        }
    } catch (e) {
        // Fallback
    }
    
    // Fallback to mac address or hostname hash
    const hash = crypto.createHash('sha256');
    hash.update(os.hostname() + os.userInfo().username);
    return hash.digest('hex');
}

class XAuth {
    constructor(appId, appSecret, baseUrl = "http://localhost:3310") {
        this.appId = appId;
        this.appSecret = appSecret;
        this.baseUrl = baseUrl.replace(/\/$/, "");
        this.sessionId = null;
        this.nonce = null;
    }

    async initialize() {
        try {
            const res = await fetch(`${this.baseUrl}/api/v1/client/initialize`, {
                method: "POST",
                headers: { "Content-Type": "application/json" },
                body: JSON.stringify({ app_id: this.appId })
            });

            if (res.ok) {
                const data = await res.json();
                this.sessionId = data.session_id;
                this.nonce = data.nonce;
                return { success: true };
            }
            return { success: false, message: await res.text() };
        } catch (e) {
            return { success: false, message: e.message };
        }
    }

    _decryptResponse(encryptedData) {
        try {
            const [ivB64, tagB64, dataB64] = encryptedData.split(":");
            
            const iv = Buffer.from(ivB64, 'base64');
            const tag = Buffer.from(tagB64, 'base64');
            const ciphertext = Buffer.from(dataB64, 'base64');

            const rawKey = Buffer.from(this.appSecret + this.nonce, 'utf-8');
            const key = crypto.createHash('sha256').update(rawKey).digest();

            const decipher = crypto.createDecipheriv('aes-256-gcm', key, iv);
            decipher.setAuthTag(tag);
            
            let decrypted = decipher.update(ciphertext);
            decrypted = Buffer.concat([decrypted, decipher.final()]);
            
            return JSON.parse(decrypted.toString('utf-8'));
        } catch (e) {
            return null;
        }
    }

    async validateLicense(licenseKey) {
        if (!this.sessionId || !this.nonce) {
            const init = await this.initialize();
            if (!init.success) {
                return { success: false, message: `Initialization failed: ${init.message}` };
            }
        }

        const hwid = getHwid();
        
        try {
            const res = await fetch(`${this.baseUrl}/api/v1/client/validate`, {
                method: "POST",
                headers: { "Content-Type": "application/json" },
                body: JSON.stringify({
                    license_key: licenseKey,
                    hwid: hwid,
                    app_secret: this.appSecret,
                    session_id: this.sessionId
                })
            });

            if (res.ok) {
                const body = await res.json();
                const decryptedData = this._decryptResponse(body.data);
                
                if (decryptedData && decryptedData.status === "success") {
                    return {
                        success: true,
                        expiry: decryptedData.expiry,
                        broadcast: decryptedData.broadcast,
                        variables: decryptedData.variables || {}
                    };
                }
                return { success: false, message: "Failed to verify secure response." };
            } else {
                try {
                    const errBody = await res.json();
                    return { success: false, message: errBody.message || "Validation rejected" };
                } catch {
                    return { success: false, message: await res.text() };
                }
            }
        } catch (e) {
            return { success: false, message: `Network error: ${e.message}` };
        }
    }

    async checkVersion(currentVersion, channel = "stable") {
        try {
            const res = await fetch(`${this.baseUrl}/api/v1/client/verify-version`, {
                method: "POST",
                headers: { "Content-Type": "application/json" },
                body: JSON.stringify({
                    app_id: this.appId,
                    app_secret: this.appSecret,
                    channel: channel,
                    current_version: currentVersion
                })
            });

            if (res.ok) {
                const data = await res.json();
                return {
                    success: true,
                    version: data.version,
                    channel: data.channel,
                    url: data.url,
                    checksum: data.checksum,
                    publishedAt: data.published_at,
                    updateAvailable: data.update_available,
                    broadcast: data.broadcast
                };
            } else {
                try {
                    const errBody = await res.json();
                    return { success: false, message: errBody.message || "Version check failed" };
                } catch {
                    return { success: false, message: await res.text() };
                }
            }
        } catch (e) {
            return { success: false, message: `Network error: ${e.message}` };
        }
    }

    async getLatestRelease(channel = "stable") {
        try {
            const res = await fetch(`${this.baseUrl}/api/update/${this.appId}/${channel}`);
            if (res.ok) {
                const data = await res.json();
                return {
                    success: true,
                    version: data.version,
                    channel: data.channel,
                    url: data.url,
                    checksum: data.checksum,
                    publishedAt: data.published_at
                };
            }
            return { success: false, message: await res.text() };
        } catch (e) {
            return { success: false, message: e.message };
        }
    }
}

// Example Usage
if (require.main === module) {
    (async () => {
        const auth = new XAuth(1, "your_app_secret_here");

        console.log("--- XAuth Omega Node.js JS SDK Demo ---");

        // 1. Public Update Check
        console.log("\n[1] Checking for updates (Public)...");
        const release = await auth.getLatestRelease();
        if (release.success) {
            console.log(`Latest Version: ${release.version} (${release.channel})`);
        }

        // 2. Secure Check
        console.log("\n[2] Verifying version (Secure)...");
        const ver = await auth.checkVersion("1.0.0");
        if (ver.success) {
            console.log(ver.updateAvailable ? "Update available!" : "Up to date.");
        }

        // 3. License Validation
        console.log("\n[3] Validating license...");
        const result = await auth.validateLicense("XXXX-XXXX-XXXX-XXXX");
        console.log(result.success ? `Valid! Expiry: ${result.expiry}` : `Failed: ${result.message}`);
    })();
}

module.exports = XAuth;
